proof/x86/ctx_inverse_mod_384.saw

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2021 Galois, Inc.
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+*/
+
+///////////////////////////////////////////////////////////////////////////////
+// Specifications
+///////////////////////////////////////////////////////////////////////////////
+
+// ct_inverse_mod_383
+let ct_inverse_mod_383_spec = do {
+  ret_ptr <- crucible_alloc vec768_type;
+  (_, inp_ptr) <- ptr_to_fresh_readonly "inp" vec384_type;
+  (_, mod_ptr) <- ptr_to_fresh_readonly "mod" vec384_type;
+  crucible_execute_func [ret_ptr, inp_ptr, mod_ptr];
+  new_ret <- crucible_fresh_var "new_ret" vec768_type;
+  crucible_points_to ret_ptr (crucible_term new_ret);
+};
+
+///////////////////////////////////////////////////////////////////////////////
+// Proofs
+///////////////////////////////////////////////////////////////////////////////
+
+// Some elements of ctx_inverse_mod_383 present some performance problems for the
+// simulator. We enable some options here in order to make things easier:
+// - enable_x86_what4_hash_consing improves term sharing during simulation in
+//   some cases.
+// - set_x86_stack_base_align sets the alignment of the base of the stack
+//   allocation to 512 bytes. In the simulator's memory model, addresses are
+//   represented as an abstract base plus an offset bitvector. Since the base is
+//   abstract, we know little about the actual value of the address, making it
+//   difficult to perform bitwise operations on addresses. However, if the base
+//   is known to have a certain alignment, we know that a certain number of low
+//   bits of the address are determined entirely by the offset. This makes it
+//   possible to apply bitwise operations that only influence those bits
+//   directly to the offset.
+
```

Comment:
It might be worth further elaborating this point.  Does it affect the soundness of the analysis?  If not, why not?

Reply:
@chameco let's mention that `set_x86_stack_base_align 512` means that `ctx_inverse_mod_383` is verified under the assumption that `%rsp` is 512-byte aligned, while the ABI requires only 16-byte alignment. This is a limitation. It would be possible to verify all 32 bit patterns, but it would take a long time.